### PR TITLE
backend: add IDrive e2 to s3 provider list

### PR DESCRIPTION
**Pull request by vyloy**
#### What is the purpose of this change?

Add IDrive e2 to s3 provider list.

#### Was the change discussed in an issue or in the forum before?

No.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: fs/config/ui.go [hunk 1]
```diff
@@ -380,6 +382,33 @@ func backendConfig(ctx context.Context, name string, m configmap.Mapper, ri *fs.
 			if out.Option.Default == nil {
 				out.Option.Default = ""
 			}
+			if out.Option.Name == "endpoint" {
```
Comment (ncw) [inline fs/config/ui.go:385]: Can you factor the contents of this out to a separate function please.

**Review by ncw — COMMENTED** [review]
This looks basically OK.

Check e34c5436608c75432b5db5bc5431a1907236f240 as an example - there are other files you need to edit.

Can you also set a set of provider quirks please? Run the integration tests to work out which quirks you need to set.

```
cd backend/s3
go test -v -verbose -remote YourS3Remote:
```

File: fs/config/ui.go (hunk 1 above)
Comment (vyloy) [reply fs/config/ui.go]: Sure, I have set the quirks it needs and refactor the contents of this out to a separate function, please review again. 

Thanks for your time.

File: docs/content/s3.md [hunk 2]
```diff
@@ -3076,6 +3079,130 @@ acl> 1
...
+Press Enter to leave empty.
+[snip]
+10 / IBM COS S3
+   \ (IBMCOS)
+11 / IDrive e2
+   \ (IDrive)
+12 / Seagate Lyve Cloud
+   \ (LyveCloud)
+[snip]
+provider> 11
```
Comment (ncw) [inline docs/content/s3.md:3126]: And here

File: docs/content/s3.md [hunk 3]
```diff
@@ -3076,6 +3079,130 @@ acl> 1
...
+Edit advanced config?
+y) Yes
+n) No (default)
+y/n> 
+
+Configuration complete.
+Options:
+- type: s3
+- provider: IDrive
+- access_key_id: dGVzdC9CUUtYUUKyeH0C
```
Comment (ncw) [inline docs/content/s3.md:3196]: I hope this is just an example access key!

File: fs/config/ui.go [hunk 4]
```diff
@@ -355,6 +358,39 @@ func newSection() {
 	fmt.Println()
 }
 
+// setDefaultEndpointValueForIDriveE2 gets user region endpoint against the Access Key details by calling the API
```
Comment (ncw) [inline fs/config/ui.go:361]: Sorry, I meant factor this into `backend/s3/s3.go` - it doesn't belong here as it is is backend specific.

File: docs/content/s3.md [hunk 5]
```diff
@@ -3076,6 +3079,130 @@ acl> 1
...
+Enter name for new remote.
+name> e2
+
+Option Storage.
+Type of storage to configure.
+Choose a number from below, or type in your own value.
+[snip]
+ 4 / Amazon Drive
+   \ (amazon cloud drive)
+ 5 / Amazon S3 Compliant Storage Providers including AWS, Alibaba, Ceph, China Mobile, Cloudflare, ArvanCloud, Digital Ocean, Dreamhost, Huawei OBS, IBM COS, IDrive e2, Lyve Cloud, Minio, Netease, RackCorp, Scaleway, SeaweedFS, StackPath, Storj, Tencent COS and Wasabi
```
[the remark this answers is not in the document]
Comment (vyloy) [reply docs/content/s3.md]: Got it.

File: fs/config/ui.go (hunk 4 above)
Comment (vyloy) [reply fs/config/ui.go]: Got it.

File: fs/config/ui.go [hunk 6]
```diff
@@ -380,6 +381,7 @@ func backendConfig(ctx context.Context, name string, m configmap.Mapper, ri *fs.
 			if out.Option.Default == nil {
 				out.Option.Default = ""
 			}
+			s3.SetDefaultEndpointValueForIDriveE2(out, m)
```
Comment (ncw) [inline fs/config/ui.go:384]: Can you move this call into `backend/s3/s3.go` too please?

Comment (vyloy) [reply fs/config/ui.go:384]: Please advise where is the right place to be moved to?

Comment (ncw) [reply fs/config/ui.go:384]: Sorry, I misunderstood what your code was doing.

You need to call this in the `Config` callback which is called after all the initial config is done. The S3 backend hasn't got one, so you need to add one something like this:

```diff
 		Description: "Amazon S3 Compliant Storage Providers including AWS, Alibaba, Ceph, China Mobile, Cloudflare, ArvanCloud, Digital Ocean, Dreamhost, Huawei OBS, IBM COS, Lyve Cloud, Minio, Netease, RackCorp, Scaleway, SeaweedFS, StackPath, Storj, Tencent COS and Wasabi",
 		NewFs:       NewFs,
 		CommandHelp: commandHelp,
+		Config: func(ctx context.Context, name string, m configmap.Mapper, config fs.ConfigIn) (*fs.ConfigOut, error) {
+			switch config.State {
+			case "":
+				return nil, setDefaultEndpointValueForIDriveE2(ctx, m)
+			}
+			return nil, fmt.Errorf("unknown state %q", config.State)
+		},
 		Options: []fs.Option{{
 			Name: fs.ConfigProvider,
 			Help: "Choose your S3 provider.",
```

You can set values in the `m` at that point using `m.Set("endpoint", newValue)`.

You may want to stop the `endpoint` question being asked at all.

You can also ask the user questions too, but that requires adding more states.

Hopefully that is clear - let me know if it isn't and sorry for the confusion.

Comment (vyloy) [reply fs/config/ui.go]: Thank you very much for your help and patient!

**Comment by vyloy** [comment]
@ncw Hi, pushed an update, please review.

**Review by ncw — APPROVED** [review]
That looks perfect - thank you.

I'll just run the CI and if it passes - I'll merge - thank you very much :-)

**Comment by ncw** [comment]
I made a small fix for the CI problem - hopefully it will go green and I can merge this time :-)

**Comment by ncw** [comment]
Thank you :-)

**Pull request merged**